File: rust/guessthenumber/src/game.rs
// Internal modules
mod evaluator;
mod common;

use evaluator::Answer;
use evaluator::evaluate_guess;


pub trait Game {
    fn play(self) -> std::result::Result<(), Box<dyn std::error::Error>>;
}

pub struct HumanGuessesMachineThinksGame;

impl Game for HumanGuessesMachineThinksGame {
    fn play(self) -> std::result::Result<(), Box<dyn std::error::Error>> {
        let thought_number = common::generate_stringified_random_number(4);
        println!("Thought number: {}", thought_number);
        let mut guess = ask_for_a_guess();
        let mut answer: Answer = evaluate_guess(&guess, &thought_number);

        loop {
            if answer.is_correct(&guess) {
                break;
            } else {
                println!("{:?}. Try again!", answer);
                guess = ask_for_a_guess();
                answer = evaluate_guess(&guess, &thought_number);
            }
        }

        println!("You win!");

        Ok(())
    }
}

fn ask_for_a_guess() -> String {
    try_read!().unwrap()
}

File: python/guessthenumber/game.py
from abc import (
    ABC,
    abstractmethod,
)
from typing import Optional

from guessthenumber.constants import GAME_OVER_MESSAGE
from guessthenumber.evaluator import (
    Answer,
    Evaluator,
)
from guessthenumber.guesser import Guesser
from guessthenumber.thinker import Thinker


class Game(ABC):

    @abstractmethod
    def play(self, cipher_quantity: int = 4, additional_message: str = None) -> None:
        raise NotImplementedError('Must implement a subclass.')


class MachineGuessesHumanThinksGame(Game):

    def play(self, cipher_quantity: int = 4, additional_message: str = None) -> None:

        guesser: Guesser = Guesser(cipher_quantity)
        hint: Optional[str] = None

        # import pdbpp; pdbpp.set_trace()
        while True:
            guess: Optional[str] = guesser.take_a_guess(hint)

            if guess is None:
                break

            hint = input(f'My guess: {guess}. If not right, please give a hint:\n--> ')

        print(GAME_OVER_MESSAGE)


class HumanGuessesMachineThinksGame(Game):

    # TODO: cipher_quantity is not used but at some point the machine
    # should ask how many ciphers are going to be used to play.
    def play(self, cipher_quantity: int = 4, additional_message: str = None) -> None:
        number: str = Thinker.think_number(cipher_quantity=4)

        while True:
            guess: str = self.ask_for_a_guess(additional_message)
            answer: Answer = Evaluator.evaluate_guess(guess, number)

            if answer:
                break

            additional_message = str(answer)

        print(GAME_OVER_MESSAGE)

    @staticmethod
    def ask_for_a_guess(additional_message: str = None) -> str:
        if additional_message:
            print(additional_message)

        return input('Please, take a guess: ')


if __name__ == '__main__':
    which_game: int = int(
        input(
            'Hey there! There are two games to play here:\n\t'
            '1. Machine guesses, you think\n\t'
            '2. You guess, machine thinks\n'
            '-->  '
        )
    )
    if which_game == 1:
        game: Game = MachineGuessesHumanThinksGame()
    elif which_game == 2:
        game = HumanGuessesMachineThinksGame()
    else:
        print('Hmm, sorry! There\'s no such game.')

    game.play()
